cmd/thanos/bucket.go

```diff
@@ -291,6 +299,123 @@ func registerBucketInspect(m map[string]setupFunc, root *kingpin.CmdClause, name
+		ctx, cancel := context.WithCancel(context.Background())
+
+		router := route.New()
+
+		bucketUI := ui.NewBucketUI(logger)
+		bucketUI.Register(router)
+
+		g.Add(func() error {
+			return refresh(ctx, logger, bucketUI, int(*interval), name, reg, objStoreConfig)
+		}, func(error) {
+			cancel()
+		})
+
+		l, err := net.Listen("tcp", *bind)
+		if err != nil {
+			return errors.Wrapf(err, "listen HTTP on address %s", *bind)
+		}
+
+		g.Add(func() error {
+			level.Info(logger).Log("msg", "Listening for query and metrics", "address", *bind)
+			return errors.Wrap(http.Serve(l, router), "serve web")
+		}, func(error) {
+			runutil.CloseWithLogOnErr(logger, l, "http listener")
+		})
+
+		return nil
+	}
+}
+
+// Refresh metadata from remote storage periodically and update UI
+func refresh(ctx context.Context, logger log.Logger, bucketUI *ui.Bucket, interval int, name string,
+	reg *prometheus.Registry, objStoreConfig *pathOrContent) error {
+
+	confContentYaml, err := objStoreConfig.Content()
+	if err != nil {
+		return err
+	}
+
+	// A client is created for the entire duration of the program to avoid duplicate metrics registration.
+	bkt, err := client.NewBucket(logger, confContentYaml, reg, name)
+	if err != nil {
+		return err
+	}
+
+	defer runutil.CloseWithLogOnErr(logger, bkt, "bucket client")
+
+	delay := time.Duration(interval) * time.Minute
+
+	return runutil.Repeat(delay, ctx.Done(), func() error {
+		iterCtx, iterCancel := context.WithTimeout(ctx, 5*time.Minute)
+		defer iterCancel()
+
+		// If there are no errors and the data isn't stale, wait a while before updating again.
+		if bucketUI.Err == nil && time.Since(bucketUI.RefreshedAt) < (5*time.Minute) {
+			return nil
+		}
+
+		// Occasional network failures are OK; retry until download succeeds. Exponential decay would work
+		// much better than a simple loop.
+		return runutil.RetryWithLog(logger, time.Minute, iterCtx.Done(), func() error {
+			blocks, err := download(iterCtx, logger, bkt)
+			if err != nil {
+				bucketUI.Set("[]", err)
+				return err
+			}
+
+			data, err := json.Marshal(blocks)
+			if err != nil {
+				bucketUI.Set("[]", err)
+				return err
+			}
+			bucketUI.Set(string(data), nil)
+			return nil
+		})
+	})
+}
+
+func download(ctx context.Context, logger log.Logger, bkt objstore.Bucket) (blocks []metadata.Meta, err error) {
+	level.Info(logger).Log("msg", "synchronizing block metadata")
+
+	err = bkt.Iter(ctx, "", func(name string) error {
+		id, ok := block.IsBlockDir(name)
+		if !ok {
+			return nil
+		}
+
+		meta, err := block.DownloadMeta(ctx, logger, bkt, id)
+		if err != nil {
+			return err
+		}
+
+		blocks = append(blocks, meta)
+		return nil
+	})
+
+	if err != nil {
+		level.Error(logger).Log("err", err, "msg", "Failed to downloaded block metadata")
+		return blocks, err
+	}
+
+	level.Info(logger).Log("msg", "downloaded block metadata")
```

Comment:
Can we add number of blocks e.g we syncronized? Also let's be explicit:

```suggestion
	level.Info(logger).Log("msg", "downloaded blocks meta.json", "num", len(blocks))
```